change to unidirectional signal called "Incremental" (#4)
kazuho/draft-kazuho-httpbis-streaming-requests: draft-kazuho-httpbis-streaming-requests.md

informative:
  PROXY-STATUS: RFC9209
  SSE:
    target: https://html.spec.whatwg.org/multipage/server-sent-events.html
    title: Server-Sent Events
    author:
     -
        org: WHATWG

--- abstract

This document specifies the "Incremental" HTTP header field, which instructs
HTTP intermediaries to forward the HTTP message incrementally.

--- middle

# Introduction

HTTP {{!HTTP=RFC9110}} permits receivers to begin processing portions of HTTP
messages as they arrive, rather than requiring them to wait for the entire HTTP
message to be received before acting.

Some applications are specifically designed to take advantage of this
capability.

For example, Server-Sent Events {{SSE}} uses a long-running HTTP response, where
the server continually sends notifications as they become available.

In the case of Chunked Oblivious HTTP Messages
{{?CHUNKED-OHTTP=I-D.ietf-ohai-chunked-ohttp}}, the client opens an HTTP request

> and incrementally sends application messages, while the server begins responding

Reviewer: ```suggestion
and incrementally sends application messages, while the server can start responding
```

even before the HTTP request is fully complete. In this way, the HTTP
request-response pair effectively serves as a bi-directional communication
channel.

However, these applications are fragile when HTTP intermediaries are involved.
This is because HTTP intermediaries are not only permitted but are frequently
deployed to buffer complete HTTP messages before forwarding them downstream
({{Section 7.6 of HTTP}}).

If such a buffering HTTP intermediary exists between the client and the server,
these applications may fail to function as intended.

In the case of Server-Sent Events, when an intermediary tries to buffer the HTTP

> response completely before forwarding it, the client can time out before

Reviewer 2: ```suggestion
response completely before forwarding it, the client might time out before
```

receiving any portion of the HTTP response.

In the case of Chunked Oblivious HTTP Messages, when an intermediary tries to
buffer the entire HTTP request, the client will not start receiving application
messages from the server until the client closes the request, effectively

> disrupting the intended bi-directional communication.

Reviewer 2: Doesn't necessarily need to be bi-directional...
```suggestion
disrupting the intended incremental processing of the request.
```

> To prevent such deadlocks, this document specifies the "Incremental" HTTP header

Reviewer 2: ```suggestion
To prevent such behavior, this document specifies the "Incremental" HTTP header
```

> To prevent such behavior, this document specifies the "Incremental" HTTP header

Reviewer: ```suggestion
To help avoid such behavior, this document specifies the "Incremental" HTTP header
```

The term Boolean is imported from {{!STRUCTURED-FIELDS=RFC8941}}.

# The Incremental Header Field

The Incremental HTTP header field expresses the sender's intent for HTTP
intermediaries to start forwarding the message downstream before the entire
message is received.

This header field has just one valid value of type Boolean: "?1".

~~~
Incremental = ?1
~~~

Upon receiving a header section that includes the Incremental header field, HTTP
intermediaries SHOULD NOT buffer the entire message before forwarding it.
Instead, intermediaries SHOULD transmit the header section downstream and
continuously forward the bytes of the message body as they arrive.

The Incremental HTTP header field applies to each HTTP message. Therefore, if
both the HTTP request and response need to be forwarded incrementally, the
Incremental HTTP header field MUST be set for both the HTTP request and the
response.

> 

Reviewer: ```suggestion

The Incremental field is advisory. Intermediaries that are unaware of the field
or that do not support the field might buffer messages, even when explicitly
requested otherwise.  Clients and servers therefore cannot expect all
intermediaries to understand and respect a request to deliver messages
incrementally. Clients can rely on prior knowledge or probe for support on
individual resources.
```

(Happy to take that to a separate change; it's not exactly simple.)

Author: Let's get this in. As you say we might want to discuss, I think it is a fact that the signal is only advisory; it makes sense to point that out.

HTTP requests that they forward, while buffering requests that exceed this
limit.

> Such intermediaries may apply a more restrictive concurrency limit to requests

Reviewer: ```suggestion
Such intermediaries could apply a more restrictive concurrency limit to requests
```
